File: python/Backtracking/palindrome_partitioning.py
from typing import List

class Solution:
    def partition(self, s: str) -> List[List[str]]:
        """
        Approach:
        - Use backtracking (DFS) to explore all possible partitions of the string.
        - At each step, try every possible substring starting at the current index.
        - If the substring is a palindrome, add it to the current partition and recurse.
        - If the end of the string is reached, add the current partition to the result.
        - Use a helper function to check if a substring is a palindrome.

        Time Complexity: O(2^n * n), where n is the length of s (there are 2^n possible partitions, and each palindrome check is O(n)).
        Space Complexity: O(n) for the recursion stack, plus O(2^n * n) for the result storage.
        """
        res = []
        part = []

        def dfs(i):
            if i >= len(s):
                res.append(part.copy())
                return
            for j in range(i, len(s)):
                if self.isPali(s, i, j):
                    part.append(s[i : j + 1])
                    dfs(j + 1)
                    part.pop()
        dfs(0)
        return res
    
    def isPali(self, s, l, r):
        while l < r:
            if s[l] != s[r]:
                return False
            l, r = l + 1, r - 1
        return True
    
# -------------------
# Test Cases
# -------------------
def test_palindrome_partitioning():
    sol = Solution()
    # Test case 1: Example from Leetcode
    s1 = "aab"
    result1 = sol.partition(s1)
    expected1 = [["a","a","b"],["aa","b"]]
    assert sorted([sorted(x) for x in result1]) == sorted([sorted(x) for x in expected1])
    # Test case 2: Single character
    s2 = "a"
    assert sol.partition(s2) == [["a"]]
    # Test case 3: All palindromes
    s3 = "aaa"
    result3 = sol.partition(s3)
    expected3 = [["a","a","a"],["a","aa"],["aa","a"],["aaa"]]
    assert sorted([sorted(x) for x in result3]) == sorted([sorted(x) for x in expected3])
    # Test case 4: No palindromic partition except single letters
    s4 = "abc"
    assert sol.partition(s4) == [["a","b","c"]]
    # Test case 5: Empty string
    s5 = ""
    assert sol.partition(s5) == [[]]
    print("All test cases passed!")

if __name__ == "__main__":
    test_palindrome_partitioning()
    

File: C++/Backtracking/palindrome_partitioning.cpp
#include <bits/stdc++.h>
using namespace std;

/**
 * Approach:
 * - Use backtracking (DFS) to explore all possible partitions of the string.
 * - At each step, iterate through all substrings starting from the current index.
 * - If the substring is a palindrome, include it in the current path and recurse.
 * - When we reach the end of the string, save the current path to the result.
 * - Use a helper function to check if a given substring is a palindrome.
 *
 * Time Complexity: O(2^n * n), where n = length of string
 *   → 2^n choices of partitions × O(n) to check palindrome
 * Space Complexity: O(n) (recursion stack) + O(2^n * n) (result)
 */

class Solution {
public:
    vector<vector<string>> res;      // To store all valid partitions
    vector<string> part;            // Current partition being built

    vector<vector<string>> partition(string s) {
        dfs(0, s);
        return res;
    }

private:
    void dfs(int i, string& s) {
        if (i >= s.size()) {
            res.push_back(part);    // Reached end, save current partition
            return;
        }
        for (int j = i; j < s.size(); ++j) {
            if (isPalindrome(s, i, j)) {
                part.push_back(s.substr(i, j - i + 1));  // choose substring
                dfs(j + 1, s);                           // recurse
                part.pop_back();                         // backtrack
            }
        }
    }

    bool isPalindrome(string& s, int l, int r) {
        while (l < r) {
            if (s[l] != s[r]) return false;
            ++l;
            --r;
        }
        return true;
    }
};

/*
------------------------------------------------------
📄 Line-by-Line C++ Syntax Explanation:
------------------------------------------------------

#include <bits/stdc++.h>
    → Includes all standard C++ headers (vector, string, etc.)

using namespace std;
    → Avoids having to prefix everything with std::

class Solution { ... };
    → Defines a class named Solution.

vector<vector<string>> res;
    → A 2D vector to store all the partitions.

vector<string> part;
    → Stores the current path (one possible partition).

vector<vector<string>> partition(string s)
    → Public function that starts the process, returns all partitions.

dfs(0, s);
    → Start DFS at index 0 with the string s.

void dfs(int i, string& s)
    → Recursive helper function to explore partitions starting at index i.

if (i >= s.size())
    → Base case: reached end of string.

res.push_back(part);
    → Save a copy of current partition.

for (int j = i; j < s.size(); ++j)
    → Try all substrings starting at i and ending at j.

if (isPalindrome(s, i, j))
    → If current substring is a palindrome, proceed.

part.push_back(s.substr(i, j - i + 1));
    → Add substring [i..j] to the current partition.

dfs(j + 1, s);
    → Recurse from the next index after j.

part.pop_back();
    → Backtrack: remove last added substring.

bool isPalindrome(string& s, int l, int r)
    → Checks if substring s[l..r] is a palindrome.

while (l < r) { ... }
    → Compare characters from both ends toward the center.

return true;
    → If all characters match, it is a palindrome.

------------------------------------------------------

⏱ Complexity:
- Time: O(2^n × n), exponential because of all possible partitions.
- Space: O(n) for recursion stack + result storage.
------------------------------------------------------
*/

